Slim/Control/Queries.pm

```diff
@@ -905,21 +884,36 @@ sub albumsQuery {
 					$contributorSth->bind_param(":work", $work);
 					$contributorSth->bind_param(":performance", $c->{'tracks.performance'}||undef);
 				}
-				$contributorSth->execute();
+				my $contributorArray = $dbh->selectall_arrayref($contributorSth,{ Slice => {} });
+
+				my $contributorHash = {};
+				foreach (@$contributorArray) {
+					push @{$contributorHash->{$_->{'role'}}->{'id'}}, $_->{'id'};
+					push @{$contributorHash->{$_->{'role'}}->{'name'}}, $_->{'name'};
+				}
 
-				my $contributor = $contributorSth->fetchrow_hashref;
-				$contributorSth->finish;
+				my @artists;
+				my @artistIds;
+				foreach my $role ( map { Slim::Schema::Contributor->typeToRole($_) } @linkRoles ) {
+					foreach my $name ( @{$contributorHash->{$role}->{'name'}} ) {
+						push @artists, $name unless grep(/$name/, @artists);
+					}
+					foreach my $id ( @{$contributorHash->{$role}->{'id'}} ) {
+						push @artistIds, $id unless grep(/$id/, @artistIds);
+					}
+				}
```

Comment:
And finally we create the output arrays, removing duplicates:
```

@artists=(
  "Gert Mortensen",
  "Kjell-Ake Pettersson",
  "Malm\xC3\xB6 Symphony Orchestra",
  "Per Falck",
  "Okko Kamu",
)

@artistIds=(530_486, 530_481, 530_483, 530_482, 530_485)

```

Reply:
Using the `@linkRoles` array for selection and sequencing:

`@linkRoles=("ARTIST", "ALBUMARTIST", "CONDUCTOR", "BAND", "TRACKARTIST")`

Reply:
We could save a few CPU cycles here: there's no need to map `@linkRoles` over and over again, is there? That list doesn't change inside the loop. We seem to only handle the numerical value, yet we build the text list. Even in the above statements outside the loop we seem to always convert to numbers.

And a `grep` inside a loop inside a loop inside another loop is screaming at me 😆. I'd rather push all values on `@artists` and `@artistIds`, then use on single `@artists = List::Utils::uniq(@artists)`. Unless that completely untested code is completely broken. Don't know why we never use that function. We already import some other functions from `List::Util`. This could potentially be faster than the `%seen` trick used above, as there's a XS (native C) implementation.

Reply:
Just pushed the changes, including what I hope is a further optimisation.